libtransact/src/context/manager.rs

```diff
@@ -17,66 +17,34 @@
 
 use std::collections::HashMap;
 use std::collections::VecDeque;
-use std::error::Error as StdError;
+use std::marker::Send;
 use std::str;
 
-use crate::context::{Context, ContextId};
-use crate::receipts::{
-    Event, StateChange, TransactionReceipt, TransactionReceiptBuilder,
-    TransactionReceiptBuilderError,
-};
-use crate::state::error::StateReadError;
+use super::{error::ContextManagerError, Context, ContextId, ContextLifecycle};
+use crate::receipts::{Event, StateChange, TransactionReceipt, TransactionReceiptBuilder};
 use crate::state::Read;
 
-#[derive(Debug)]
-pub enum ContextManagerError {
-    MissingContextError(String),
-    TransactionReceiptBuilderError(TransactionReceiptBuilderError),
-    StateReadError(StateReadError),
-}
-
-impl StdError for ContextManagerError {
-    fn description(&self) -> &str {
-        match *self {
-            ContextManagerError::MissingContextError(ref msg) => msg,
-            ContextManagerError::TransactionReceiptBuilderError(ref err) => err.description(),
-            ContextManagerError::StateReadError(ref err) => err.description(),
-        }
-    }
-}
-
-impl std::fmt::Display for ContextManagerError {
-    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
-        match *self {
-            ContextManagerError::MissingContextError(ref s) => {
-                write!(f, "Unable to find specified Context: {:?}", s)
-            }
-            ContextManagerError::TransactionReceiptBuilderError(ref err) => {
-                write!(f, "A TransactionReceiptBuilder error occured: {}", err)
-            }
-            ContextManagerError::StateReadError(ref err) => {
-                write!(f, "A State Read error occured: {}", err)
-            }
-        }
-    }
+pub struct ContextManager<R: Read<StateId = String, Key = String, Value = Vec<u8>>> {
+    contexts: HashMap<ContextId, Context>,
+    database: R,
 }
 
-impl From<TransactionReceiptBuilderError> for ContextManagerError {
-    fn from(err: TransactionReceiptBuilderError) -> Self {
-        ContextManagerError::TransactionReceiptBuilderError(err)
+impl<R> ContextLifecycle<R> for ContextManager<R>
```

Comment:
We need to remove R here I think; this is not something the scheduler would/should know about.